src/test/APIConsumer.t.sol

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+import {DSTest} from "ds-test/test.sol";
+import {stdError, stdCheats} from "forge-std/stdlib.sol";
+import {Vm} from "forge-std/Vm.sol";
+import {APIConsumer} from "../APIConsumer.sol";
+import {LinkToken} from "./mocks/LinkToken.sol";
+import {MockOracle} from "./mocks/MockOracle.sol";
+
+contract APIConsumerTestBase is DSTest, stdCheats {
+    Vm public constant vm = Vm(HEVM_ADDRESS);
+
+    LinkToken linkToken;
+    APIConsumer apiConsumer;
+    MockOracle mockOracle;
+
+    bytes32 public requestId;
+    uint256 public callbackValue = 777;
+
+    function setUp() public {
+        linkToken = new LinkToken();
+        mockOracle = new MockOracle(address(linkToken));
+        apiConsumer = new APIConsumer(
+            address(mockOracle),
+            "d5270d1c311941d0b08bead21fea7747",
+            10**17,
+            address(linkToken)
+        );
+        requestId = keccak256(
+            abi.encodePacked(address(apiConsumer), uint256(1))
+        );
+        linkToken.transfer(address(apiConsumer), 1000 ether);
+    }
+}
+
+contract APIConsumerUnitTest is APIConsumerTestBase {
+    event ChainlinkRequested(bytes32 indexed id);
+    event ChainlinkFulfilled(bytes32 indexed id);
+
+    /** Gas benchmarking
+     *************************************************************************/
+
+    function testGas_apiRequest() public {
+        apiConsumer.requestVolumeData();
+    }
+
+    /** Correctness tests
+     *************************************************************************/
+
+    function test_apiRequest() public {
+        vm.expectEmit(true, false, false, true);
+        emit ChainlinkRequested(requestId);
```

Comment:
Could you explain what this is doing?

Reply:
Sure, so the `expectEmit` cheatcode is used to prepare an expected log with `(bool checkTopic1, bool checkTopic2, bool checkTopic3, bool checkData)` specifying which topic(s) to check, and the calldata. We then emit the event we expect to see followed by a call to the originating function.

After the call there are internal checks for if logs were emitted in the expected order with the expected topics and data (as specified by the booleans). See Foundry book [here](https://onbjerg.github.io/foundry-book/reference/cheatcodes.html?highlight=expectem#expectemit) for a complete explanation.

So in this case we are checking whether the `ChainlinkRequested` event was emitted with the correct `requestId` on call to `requestVolumeData`. If we wanted to test event emission only then it would be something like this:

```
vm.expectEmit(false, false, false, false);
emit ChainlinkRequested(requestId);
apiConsumer.requestVolumeData();
```